test/reduce.js

```diff
@@ -128,11 +125,42 @@ module.exports = function reduce_test(testcase, minify_options, reduce_options)
                 var expr = [
                     node.expression,
                     node.args[0],
-                ][ node.start._permute++ % 2 ];
+                    null,  // intentional
+                ][ ((node.start._permute += step) * steps | 0) % 3 ];
                 if (expr) {
                     CHANGED = true;
                     return expr;
                 }
+                if (node.expression instanceof U.AST_Function) {
+                    // hoist and return expressions from the IIFE function expression
+                    var body = node.expression.body;
+                    node.expression.body = [];
+                    var seq = [];
+                    body.forEach(function(node) {
+                        var expr = expr instanceof U.AST_Exit ? node.value : node.body;
+                        if (expr instanceof U.AST_Node && !is_statement(expr)) {
+                            // collect expressions from each statements' body
+                            seq.push(expr);
+                        }
+                    });
+                    CHANGED = true;
+                    return to_sequence(seq);
+                }
+            }
+            else if (node instanceof U.AST_Defun) {
+                switch (((node.start._permute += step) * steps | 0) % 2) {
+                  case 0:
+                    CHANGED = true;
+                    return List.skip;
+                  case 1:
+                    if (!has_exit(node)) {
+                        var body = node.body;
+                        node.body = [];
+                        body.push(node);
+                        CHANGED = true;
+                        return List.splice(body);
```

Comment:
The function declaration with a now empty `body` was pushed after its original (now hoisted) `body` to prevent the future toplevel function call from failing. These calls do not have `typeof` checks. The empty function decl and the now useless call will be dropped in the next pass.